[REVIEW] Row-wise reduction operations via CuPy

beckernick opened the pull request:
**Summary of Changes**
This PR implements the following row-wise operations by leveraging CuPy via the cuda array interface:
- max
- min
- sum
- product
- mean
- var
- std
~- cumsum~
~- cumprod~

This is a follow-up to the discussion in #2416 , and can serve as stopgap functionality until these kinds of operations are supported by libcudf.

This implementation does not support columns with nulls (due to relying on `as_gpu_matrix`). But, if there are `NaNs` and `nan_as_null=False`, the result for rows containing `NaN` will be `null`. This may be worth adding a warning, but I suspect this is not a common case.

~If CuPy is not available, row-wise operations are not available and an ImportError is raised indicating that the user should install cupy if they need these operations.~

CuPy should be available now that it's a required dependency, so we could take this opportunity to update `.values` and remove the CuPy import checking logic.

beckernick commented:
There are a couple of in-progress PRs that optionally depend on CuPy. Both implement a helper function to do a logical check to see if CuPy is available, but they naturally differ slightly. We should standardize on a single way. @brandon-b-miller , @kkraus14 do you have any preferences on naming conventions / specific exception to catch?

beckernick commented:
rerun tests

brandon-b-miller commented:
@beckernick I was wondering about this too - the way you have it seems to be the most concise, however I assumed we'd end up merging this code out by adding CuPy as a direct dependency at some point. From that perspective I think whatever "works" for now and keeps the code functioning intuitively is fine.

beckernick commented:
>I assumed we'd end up merging this code out by adding CuPy as a direct dependency at some point. From that perspective I think whatever "works" for now and keeps the code functioning intuitively is fine.

Agreed. In terms of the code, I'm happy either way. Perhaps whichever is merged first can just be the temporary approach unless Keith has a preference

beckernick commented:
This is now ready for review

beckernick commented:
This is now ready for review

kkraus14 commented:
I'm not comfortable trying to squeeze this into 0.10, so pushing to 0.11

kkraus14 commented on python/cudf/cudf/core/dataframe.py:
```diff
@@ -3528,66 +3528,144 @@ def quantile(
     # Stats
     #
+    def _as_common_dtype(self):
+        common_dtype = np.find_common_type(self.dtypes, [])
```
This doesn't handle datetime dtypes, should use `np.result_type(*self.dtypes)` instead.

kkraus14 commented on python/cudf/cudf/core/dataframe.py:
```diff
@@ -3528,66 +3528,144 @@ def quantile(
+            raise ValueError(msg)
+
+        coerced = self._as_common_dtype()
+        arr = cp.asarray(coerced.as_gpu_matrix())
```
This will return a column-major matrix, where you likely want row-major.
```suggestion
        arr = cp.asarray(coerced.as_gpu_matrix(order='C'))
```

kkraus14 commented on python/cudf/cudf/core/dataframe.py:
```diff
@@ -3528,66 +3528,144 @@ def quantile(
-    def min(self, **kwargs):
+    def min(self, axis=0, numeric_only=None, **kwargs):
+        if axis == 1:
+            return self._apply_rowwise_op("min", numeric_only=numeric_only)
```
This should refactored into the `_apply_support_method` function or a new function since this pattern is being repeated across numerous methods.

beckernick replied:
Refactoring this PR today, so should have a clean, reviewable PR soon.

beckernick commented:
Refactored implementation ready for review. Raises a warning in `_prepare_for_rowwise_op` that this only handles float, int, and bool dtypes rather explicitly managing the  `numeric_only` argument down the call stack and then catching the error and presenting the same warning , which is a bit hairy given our use of kwargs and cupy operations.

Also updates `df.astype` to pass arguments by keyword, rather than position.

I also removed the CuPy import checking logic in this PR, since CuPy is now a dependency. @kkraus14 , do you want me to update `.values` similarly in this PR?

beckernick commented:
Dask compatibility issues were occurring due to metadata checking requiring the underlying reduction operator to accept `skipna` as a keyword argument. Because CuPy reduction ops do not accept this, we need to do one of the following if we want to keep the dispatching logic inside `_apply_support_method` (which is desirable):

- Remove `skipna` from the `kwargs` (if it exists) at the point of the call stack at which the CuPy operator is being invoked since it has no relevance.
- Change CuPy to allow the skipna argument
- Remove the ability to pass keyword arguments to CuPy reduction operators

I went with option one, since it can be done with a single line:

```python
# for dask metadata compatibility
kwargs.pop("skipna", None)
```

kkraus14 commented on python/cudf/cudf/core/dataframe.py:
```diff
@@ -3748,20 +3772,41 @@ def any(self, bool_only=None, **kwargs):
+
+        elif axis == 1:
+            # for dask metadata compatibility
+            kwargs.pop("skipna", None)
```
Before doing this can we assert that the value of `skipna` is something where this won't cause incorrect results? I.E. if `skipna` is false we throw an exception, if `skipna` is true we do this (I think that's the correct behavior, please correct me if I'm wrong).

beckernick replied:
Agreed. Very good point.

beckernick replied:
Adding the `skipna` check after popping the value. I don't love how this relies on skipna being based by keyword only, but I think it's fine since this is an internal method and it can be controlled reasonably well.

kkraus14 commented on python/cudf/cudf/core/dataframe.py:
```diff
@@ -3748,20 +3772,41 @@ def any(self, bool_only=None, **kwargs):
+            kwargs.pop("skipna", None)
+
+            prepared = self._prepare_for_rowwise_op()
+            arr = cp.asarray(prepared.as_gpu_matrix())
```
@brandon-b-miller are we okay with just using CuPy wherever now or are we still checking for whether it's installed or not?

brandon-b-miller replied:
@kkraus14 now that the updates to the conda `yml` files are merged it should be ok to just import and call CuPy anywhere in the code. In fact the code that does the checking should probably be removed now that we have a way to stitch the proper cupy/numba versions together.

beckernick replied:
Love it!
